src/include/miopen/graphapi/graphapi_variant_pack.hpp

```diff
@@ -0,0 +1,209 @@
+    void* mWorkspace = nullptr;
+
+    friend class VariantPackBuilder;
+};
+
+class VariantPackBuilder
+{
+public:
+    VariantPackBuilder& setTensorIds(const std::vector<int64_t>& tensorIds) &
+    {
+        if(!detail::noRepetitions(tensorIds))
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+        mVariantPack.mTensorIds = tensorIds;
+        mTensorIdsSet           = true;
+        return *this;
+    }
+    VariantPackBuilder& setTensorIds(std::vector<int64_t>&& tensorIds) &
+    {
+        if(!detail::noRepetitions(tensorIds))
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+        mVariantPack.mTensorIds = std::move(tensorIds);
+        mTensorIdsSet           = true;
+        return *this;
+    }
+    VariantPackBuilder& setDataPointers(const std::vector<void*>& dataPointers) &
+    {
+        if(miopen::any_of(dataPointers, [](const auto& v) { return v == nullptr; }) ||
+           !detail::noRepetitions(dataPointers))
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+        mVariantPack.mDataPointers = dataPointers;
+        mDataPointersSet           = true;
+        return *this;
+    }
+    VariantPackBuilder& setDataPointers(std::vector<void*>&& dataPointers) &
+    {
+        if(miopen::any_of(dataPointers, [](const auto& v) { return v == nullptr; }) ||
+           !detail::noRepetitions(dataPointers))
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+        mVariantPack.mDataPointers = std::move(dataPointers);
+        mDataPointersSet           = true;
+        return *this;
+    }
+    VariantPackBuilder& setWorkspace(void* workspace) &
+    {
+        if(workspace == nullptr)
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+
+        mVariantPack.mWorkspace = workspace;
+        return *this;
+    }
+
+    VariantPackBuilder&& setTensorIds(const std::vector<int64_t>& tensorIds) &&
+    {
+        return std::move(setTensorIds(tensorIds));
+    }
+    VariantPackBuilder&& setTensorIds(std::vector<int64_t>&& tensorIds) &&
+    {
+        return std::move(setTensorIds(std::move(tensorIds)));
+    }
+    VariantPackBuilder&& setDataPointers(const std::vector<void*>& dataPointers) &&
+    {
+        return std::move(setDataPointers(dataPointers));
+    }
+    VariantPackBuilder&& setDataPointers(std::vector<void*>&& dataPointers) &&
+    {
+        return std::move(setDataPointers(std::move(dataPointers)));
+    }
+    VariantPackBuilder&& setWorkspace(void* workspace) &&
+    {
+        return std::move(setWorkspace(workspace));
+    }
+
+    VariantPack build() const&
+    {
+        if(!validate())
+            MIOPEN_THROW(miopenStatusBadParm);
+        return mVariantPack;
+    }
+    VariantPack build() &&
+    {
+        if(!validate())
+            MIOPEN_THROW(miopenStatusBadParm);
+        return std::move(mVariantPack);
+    }
+
+private:
+    VariantPack mVariantPack;
```

Comment:
[Notice][not important] @amberhassaan I guess your comment is a recommendation, because the placement of data members is a matter of taste. As Rob Pike said once, "if you've chosen the right data structures and organized things well, the algorithms will almost always be self-evident". That's why I prefer to see data first, then ctor/dtor and then the rest.

Reply:
I actually agree with you. I also want the members to be put on top. It's just that with Google's silly Style Guide becoming popular, it's now in fashion to put data members at the end, which makes the least sense to me.

Reply:
[Notice] @scerzh If you read the comments above carefully, you'll see that you followed Google's Silly Style Guide, and that's not what Amber and I actually would like to see in our code)). Up to you and @amberhassaan.
![book-3-space-indentation_125x175](https://github.com/ROCm/MIOpen/assets/12697412/d27f7c82-2095-4003-ae1b-4b84eb2aea17)

Reply:
I recommend assuming this resolved, and move data members to the beginning in the future PR (that will arrive soon I guess)